12-2secondag

2secondag opened the pull request:
## 🔗 문제 링크
[스택](https://www.acmicpc.net/problem/10828)

## ✔️ 소요된 시간
1시간 반  

## ✨ 수도 코드
### ❓문제 설명 
첫째 줄에 N개의 명령의 수가 주어지고 명령에는 총 다섯 가지가 있다. 

![image](https://github.com/AlgoLeadMe/AlgoLeadMe-5/assets/128296206/712eea71-2958-4b25-987a-343b6f90c38f)

### ❓간단한 개념 설명 

**스택이란?**
가장 나중에 들어온 자료가 가장 먼저 처리되는 자료구조이다. ( Last In First Out )
![image](https://github.com/AlgoLeadMe/AlgoLeadMe-5/assets/128296206/54aa1277-11ca-4e48-9ab3-f748c9bb775e)

### ❓ 코드 설명 
```python
sen = list(map(str, sys.stdin.readline().split()))

  if sen[0] == "push":
    stack.append(int(sen[1]))

  elif sen[0] == "pop":
    if len(stack) > 0:
      print(stack[-1])
      stack.pop()
    else:
      print(-1)
```
+ 시간을 줄이기 위해서 `sys.stdin.readline()`을 사용하여 배열로 입력받은 후  배열의 첫번째 요소에 따라서 각 명령에 해당하는 작업을 수행하면 된다. 

+ ** push의 경우** 
리스트에 str형태로 저장된 문자를 int형으로 바꿔준 뒤, 스택에 저장한다. 

+ ** pop의 경우**
스택에 들어있는 정수가 있는 경우와 없는 경우로 나누어 명령을 수행한다. 


```python
elif sen[0] == "size":
    print(len(stack))

  elif sen[0] == "empty":
    if len(stack) > 0:
      print(0)
    else: 
      print(1)
```

+ **size의 경우** 
`stack.size()`를 사용하려고 하였으나 파이썬에서 스택을 리스트로 구현하는데 리스트에는 '.size()' 가 없기 때문에 길이를 알 수 있는 `len()`을 사용하여 스택에 들어있는 정수의 개수를 출력했다. 

+ **empty의 경우**
마찬가지로 `len()`을 사용하여 스택이 비어있는지 확인한다. 

```python
 elif sen[0] == "top":
    if len(stack)>0 :
      print(stack[-1])
    else:
      print(-1)
```
+ **top의 경우**
`len()`을 사용한 것과 같은 맥락으로 스택을 리스트로 구현하여 `stack.peek()`을 사용할 수 없기 때문에 `stack[-1]'을 사용하여 가장 마지막 요소를 출력한다. 
 
### ❓ 전체코드 
```python
import sys

N = int(sys.stdin.readline())
stack = []

for i in range (N):
  sen = list(map(str, sys.stdin.readline().split()))

  if sen[0] == "push":
    stack.append(int(sen[1]))

  elif sen[0] == "pop":
    if len(stack) > 0:
      print(stack[-1])
      stack.pop()
    else:
      print(-1)

  elif sen[0] == "size":
    print(len(stack))

  elif sen[0] == "empty":
    if len(stack) > 0:
      print(0)
    else: 
      print(1)

  elif sen[0] == "top":
    if len(stack)>0 :
      print(stack[-1])
    else:
      print(-1)
```

## 📚 새롭게 알게된 내용
[스택 문법 참고](https://blog.naver.com/kcw9275/223330879344)
[파이썬 입력 시간 단축 방법 ](https://velog.io/@yeseolee/Python-%ED%8C%8C%EC%9D%B4%EC%8D%AC-%EC%9E%85%EB%A0%A5-%EC%A0%95%EB%A6%ACsys.stdin.readline)

+ 반복문으로 여러줄을 입력 받아야 하는 경우에 `input()`으로 입력받게 된다면 시간 초과가 발생할 수 있다. 

해결 방안 => `sys.stdlin.readline()` 를 사용하기! 

**! 주의할 점 !**
```python
import sys
a =  sys.stdin.readline()
```
`sys.stdin.readline()`는 **한 줄 단위로 입력받기** 때문에 개행 문자가 같이 입력 받아지는 경우가 발생합니다. 또한 변수 타입이 str로 저장되기 때문에 **형병환을 거치는 작업이 필요**하다! 

-> 수정된 코드
```python
import sys
a =  int(sys.stdin.readline())
```

## 📚 궁금한 점
```python
stack = []
N = int(input())

for i in range (N):
  sen = str(input().split())
  print(sen)

  if sen[0] == 'push':
    print("push 블록 실행 확인")
    stack.append(int(sen[1]))
    
  print(stack)
```
처음에 제가 작성한 코드인데,  push 1을 입력했을 때 `if sen == "push":`문이 실행이 안되는데 왜 안 되는 걸까요?!
![image](https://github.com/AlgoLeadMe/AlgoLeadMe-5/assets/128296206/47ad0314-771e-4842-8367-43339f4e135e)

챗gpt를 돌려 정상작동하는 코드인데요... 
```python
stack = []
N = int(input())

for i in range(N):
    input_str = input().split()  # 입력 문자열을 공백을 기준으로 분리

    if input_str[0] == "push":
        print("push 블록 실행 확인")
        num = int(input_str[1])  # 숫자 부분을 추출
        stack.append(num)

    print(stack)
```
![image](https://github.com/AlgoLeadMe/AlgoLeadMe-5/assets/128296206/7d5b425f-9164-4cad-aec2-4b671623db6f)


####결론!!! 
`sen = str(input().split())` 와 `input_str = input().split()`의 차이가 뭘까욜.....?!

2secondag committed:
2024.02.15

9kyo-hwang commented:
1번 테스트 케이스에 대해 다음과 같은 코드를 실행합니다.
```py
for _ in range(int(input())):
    sen = input().split()
    print(type(sen), sen)
```
출력은 다음과 같구요.
```
<class 'list'> ['push', '1']
<class 'list'> ['push', '2']
<class 'list'> ['top']
<class 'list'> ['size']
<class 'list'> ['empty']
<class 'list'> ['pop']
<class 'list'> ['pop']
<class 'list'> ['pop']
<class 'list'> ['size']
<class 'list'> ['empty']
<class 'list'> ['pop']
<class 'list'> ['push', '3']
<class 'list'> ['empty']
<class 'list'> ['top']
```
각각의 sen은 문자열을 담고 있는 "리스트" 타입인 것을 알 수 있죠? 그래서 ```sen[0] == 'push'```와 같이 인덱스를 이용해 원소를 접근할 수 있는 거구요.

한 번 다음과 같이 변경해볼까요?
```py
for _ in range(int(input())):
    sen = str(input().split())  # string으로 타입 캐스팅
    print(type(sen), sen)
```
```
<class 'str'> ['push', '1']
<class 'str'> ['push', '2']
<class 'str'> ['top']
<class 'str'> ['size']
<class 'str'> ['empty']
<class 'str'> ['pop']
<class 'str'> ['pop']
<class 'str'> ['pop']
<class 'str'> ['size']
<class 'str'> ['empty']
<class 'str'> ['pop']
<class 'str'> ['push', '3']
<class 'str'> ['empty']
<class 'str'> ['top']
```
생긴건 똑같지만, 각각의 sen이 **문자열**로 변경되었습니다. ```['push', '1']```를 좀 더 자세히 확인해볼까요?
```py
for _ in range(int(input())):
    sen = str(input().split())
    print(type(sen), sen)
    for ch in sen:
        print(ch)
```
```
<class 'str'> ['push', '1']
[
'
p
u
s
h
'
,
 
'
1
'
]
```
즉 ```['push', '1']```이 생김새 그대로 **문자열**로 변경된 것을 알 수 있습니다. 대괄호([])조차 '문자'인 것이죠.
그렇기 때문에 위 예시의 경우 ```sen[0] == '['```인 것입니다. 리스트처럼 'push'가 아닌 거죠.

2secondag commented:
> 1번 테스트 케이스에 대해 다음과 같은 코드를 실행합니다.
> 
> ```python
> for _ in range(int(input())):
>     sen = input().split()
>     print(type(sen), sen)
> ```
> 
> 출력은 다음과 같구요.
> 
> ```
> <class 'list'> ['push', '1']
> <class 'list'> ['push', '2']
> <class 'list'> ['top']
> <class 'list'> ['size']
> <class 'list'> ['empty']
> <class 'list'> ['pop']
> <class 'list'> ['pop']
> <class 'list'> ['pop']
> <class 'list'> ['size']
> <class 'list'> ['empty']
> <class 'list'> ['pop']
> <class 'list'> ['push', '3']
> <class 'list'> ['empty']
> <class 'list'> ['top']
> ```
> 
> 각각의 sen은 문자열을 담고 있는 "리스트" 타입인 것을 알 수 있죠? 그래서 `sen[0] == 'push'`와 같이 인덱스를 이용해 원소를 접근할 수 있는 거구요.
> 
> 한 번 다음과 같이 변경해볼까요?
> 
> ```python
> for _ in range(int(input())):
>     sen = str(input().split())  # string으로 타입 캐스팅
>     print(type(sen), sen)
> ```
> 
> ```
> <class 'str'> ['push', '1']
> <class 'str'> ['push', '2']
> <class 'str'> ['top']
> <class 'str'> ['size']
> <class 'str'> ['empty']
> <class 'str'> ['pop']
> <class 'str'> ['pop']
> <class 'str'> ['pop']
> <class 'str'> ['size']
> <class 'str'> ['empty']
> <class 'str'> ['pop']
> <class 'str'> ['push', '3']
> <class 'str'> ['empty']
> <class 'str'> ['top']
> ```
> 
> 생긴건 똑같지만, 각각의 sen이 **문자열**로 변경되었습니다. `['push', '1']`를 좀 더 자세히 확인해볼까요?
> 
> ```python
> for _ in range(int(input())):
>     sen = str(input().split())
>     print(type(sen), sen)
>     for ch in sen:
>         print(ch)
> ```
> 
> ```
> <class 'str'> ['push', '1']
> [
> '
> p
> u
> s
> h
> '
> ,
>  
> '
> 1
> '
> ]
> ```
> 
> 즉 `['push', '1']`이 생김새 그대로 **문자열**로 변경된 것을 알 수 있습니다. 대괄호([])조차 '문자'인 것이죠. 그렇기 때문에 위 예시의 경우 `sen[0] == '['`인 것입니다. 리스트처럼 'push'가 아닌 거죠.

와우. 친절한 답변 아주 감사합니다.  완전 이해했어요.

miniron-v commented:
킹갓교황

miniron-v reviewed (approved):
스택은 자주 쓰이는 만큼 구현해보는 것도 좋은 경험이죠~ 백준에 이런 문제가 있는 줄은 몰랐네요. (그리고 파이썬.. 늘 느끼지만 참 편하네요...)

코테엔 문제가 없지만, 호옥시나 스택을 직접 구현해야 하는 상황이 온다면 **예외 처리**를 꼭 해줘야 합니다. 예를 들어 입력으로 `push 1 word`같은 게 온다거나... 이럴 때 어떻게 처리할 건지(잘못된 입력 에러를 띄울지, 첫 글자만 넣어줄 지, 다 넣어줄 지 등) 정해두시면 프로그램 개발 중에 예기치 못한 오류를 만나진 않을 거예요.

그래도 훌륭하게 구현해주셨네요. 고생하셨습니다!

Redish03 commented:
> 스택은 자주 쓰이는 만큼 구현해보는 것도 좋은 경험이죠~ 백준에 이런 문제가 있는 줄은 몰랐네요. (그리고 파이썬.. 늘 느끼지만 참 편하네요...)
> 
> 코테엔 문제가 없지만, 호옥시나 스택을 직접 구현해야 하는 상황이 온다면 **예외 처리**를 꼭 해줘야 합니다. 예를 들어 입력으로 `push 1 word`같은 게 온다거나... 이럴 때 어떻게 처리할 건지(잘못된 입력 에러를 띄울지, 첫 글자만 넣어줄 지, 다 넣어줄 지 등) 정해두시면 프로그램 개발 중에 예기치 못한 오류를 만나진 않을 거예요.
> 
> 그래도 훌륭하게 구현해주셨네요. 고생하셨습니다!

예외 처리에 대한 코멘트 덧붙이자면...예를 들면 stack의 크기를 지정해줬는데 그 크기를 넘어서게 데이터를 집어넣었거나, 아니면 입력으로 `push`만 주어지고 그 뒤에 넣을 값을 안 넣는, null을 넣는 경우에 대한 처리도 해줘야 하고, 민철님 처럼 push 다음에 정수형 뿐만 아니라 string, char 등의 데이터가 들어오는 경우에 대해서 처리를 해준다거나...등등등등 예외 처리를 하면 할수록 코드는 길어질 수도 있어도 사용자 관점(알고리즘 뿐만 아니라 개발에서도)에서 되게 중요한 개념이라 알면 좋을 것 같네요..!

Redish03 reviewed (approved):
고생하셨습니다 :) 코드 상으로는 딱히 말할 건 없는 것 같네요! 

파일 명 한글로 만드는 건 비추합니다..ㅎㅎ

bomik0221 reviewed (approved):
저도 코드 관련으로 드릴 말씀은 없네요! 저도 이 문제 C++ 로 해보려다가 코드 너무 길어져서 ~~귀찮아서~~ 그만둔 적이 있는데, 다른 분들 코멘트 보니 예외 처리 등등.. 배울 점이 많은 문제 같네요. 다시 풀어봐야겠어요.... ㅎㅎ

2secondag commented:
> > 스택은 자주 쓰이는 만큼 구현해보는 것도 좋은 경험이죠~ 백준에 이런 문제가 있는 줄은 몰랐네요. (그리고 파이썬.. 늘 느끼지만 참 편하네요...)
> > 코테엔 문제가 없지만, 호옥시나 스택을 직접 구현해야 하는 상황이 온다면 **예외 처리**를 꼭 해줘야 합니다. 예를 들어 입력으로 `push 1 word`같은 게 온다거나... 이럴 때 어떻게 처리할 건지(잘못된 입력 에러를 띄울지, 첫 글자만 넣어줄 지, 다 넣어줄 지 등) 정해두시면 프로그램 개발 중에 예기치 못한 오류를 만나진 않을 거예요.
> > 그래도 훌륭하게 구현해주셨네요. 고생하셨습니다!
> 
> 예외 처리에 대한 코멘트 덧붙이자면...예를 들면 stack의 크기를 지정해줬는데 그 크기를 넘어서게 데이터를 집어넣었거나, 아니면 입력으로 `push`만 주어지고 그 뒤에 넣을 값을 안 넣는, null을 넣는 경우에 대한 처리도 해줘야 하고, 민철님 처럼 push 다음에 정수형 뿐만 아니라 string, char 등의 데이터가 들어오는 경우에 대해서 처리를 해준다거나...등등등등 예외 처리를 하면 할수록 코드는 길어질 수도 있어도 사용자 관점(알고리즘 뿐만 아니라 개발에서도)에서 되게 중요한 개념이라 알면 좋을 것 같네요..!


예외 처리에 대해서도 다시 생각해봐야겠네요~!

2secondag committed:
Merge branch 'main' into 14-2secondag
